### ISSUE TITLE
Add a delegate for OnDirectoryFinished()

### ISSUE THREAD

**Issue by iSazonov** [dotnet/runtime#1953]
## Rationale ##

In PowerShell Core repo we are trying to migrate to `FileSystemEnumerable` and we need to implement Depth parameter for Get-ChildItem cmdlet (aka `dir`). The parameter limits a depth of recursion. Current solution is to add `if (depth < 0) return;` into `ShouldRecursePredicate` and `if (depth > 0) depth--` into OnDirectoryFinished().
So we have to create custom `FileSystemEnumerator` to overload `OnDirectoryFinished()`.

The request is to add new delegate for OnDirectoryFinished() so that we can utilize the standard `FileSystemEnumerable`.

## Proposed API ##

```c#
namespace System.IO.Enumeration
{
    public class FileSystemEnumerable<TResult> : IEnumerable<TResult>
    {
        public FindPredicate? ShouldIncludePredicate { get; set; } // Existing API
        public FindPredicate? ShouldRecursePredicate { get; set; } // Existing API

        public delegate void OnDirectoryFinishedDelegate(ReadOnlySpan<char> directory); // New API
        public OnDirectoryFinishedDelegate? OnDirectoryFinished { get; set; } // New API
     }
}
```

/cc @JeremyKuhne

**Comment by JeremyKuhne** [comment]
This is not the first time I've been asked for this and it is pretty straight-forward, I think we should add this for 5.0.

**Comment by iSazonov** [comment]
For full consistency I'd  add a delegate for `ContinueOnError()` too.

**Comment by danmosemsft** [comment]
@iSazonov How would the "depth limit" scenario work with these predicates? In the implementation of ShouldRecursePredicate would it examine the FileSystemEntry and then calculate the depth vs the start of the recursion (perhaps with Path.GetRelativePath and counting the double dots?)

As with any API proposal it might be useful to write out what the user code would look like.

**Comment by iSazonov** [comment]
@danmosemsft You can see draft code in PowerShell repo https://github.com/PowerShell/PowerShell/pull/12834. The PR demonstrates well real scenario of using the API.

**Comment by danmosemsft** [comment]
@iSazonov to make the API review meetings efficient we ask that all the supporting info be in the proposal  (ideally in the top post). For API proposals this ideally includes small usage examples to help demo how the API would be used

**Comment by iSazonov** [comment]
@danmosemsft I added a code snipped from the PowerShell PR.

**Comment by danmosemsft** [comment]
Thanks, that makes it clear.

**Comment by terrajobst** [comment]
* It seems `OnDirectoryFinishedDelegate` should include the `FileSystemEntry entry`
* Should `ContinueOnErrorPredicate` provide some context? It seems one would want the container and the name the child to do anything useful for logging. If we see potential for more state, we may want to allow a struct, such as `FileSystemError` instead.
* @carlossanlop you want to sync with @JeremyKuhne on how to evolve this API

```C#
namespace System.IO.Enumeration
{
    public partial class FileSystemEnumerable<TResult> : IEnumerable<TResult>
    {
        public delegate void OnDirectoryFinishedDelegate(ref FileSystemEntry entry);
        public delegate bool ContinueOnErrorPredicate(ref FileSystemEntry parent, ReadOnlySpan<char> entry, int error);

        public OnDirectoryFinishedDelegate? OnDirectoryFinishedAction { get; set; }
        public ContinueOnErrorPredicate? ShouldContinueOnErrorPredicate { get; set; }
     }
}
```

**Comment by iSazonov** [comment]
> It seems OnDirectoryFinishedDelegate should include the FileSystemEntry entry

OnDirectoryFinished is defined as
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs#L63
and my proposal follows this. If MSFT team want enhance this - I agree.

> Should ContinueOnErrorPredicate provide some context?

ContinueOnError is defined as
https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs#L70
and my proposal follows this. If MSFT team want enhance this - I agree.

**Comment by iSazonov** [comment]
Can we continue this work? I still hope to use this in the next PowerShell milestone (LTS!).

I have updated the proposal based on previous discussion and also added a new option to EnumerationOptions to support recursion depth limit.

The current implementation uses a queue to process subdirectories. This allows us to implement the recursion depth limit in the simplest way so I don't need to invent tricky code in PowerShell. I could try to pull a PR and implement the part as only the proposal will be approved.

**Comment by carlossanlop** [comment]
> _I have updated the proposal based on previous discussion and also added a new option to EnumerationOptions to support recursion depth limit._

Thank you, @iSazonov.

> _Should ContinueOnErrorPredicate provide some context? It seems one would want the container and the name the child to do anything useful for logging. If we see potential for more state, we may want to allow a struct, such as FileSystemError instead._

The parameters of `ContinueOnErrorPredicate` were questioned in the video. The request is to consider thinking of potential future expansions of the error context information, and make sure to include the container of the filesystem item that failed. This could be achieved by adding a struct which would could hold all the relevant error information. For example, you can see that [EnumerationOptions](https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.IO.FileSystem/src/System/IO/EnumerationOptions.cs#L13) was created as a class, so that it can be expanded easily without having to modify the methods that consume it as parameter.

Another thing that was mentioned is that we need to make sure the reported error makes sense across platforms. Windows does not throw the same errors as Linux, but we need to reflect that in our APIs in an OS independent way.

> _For API proposals this ideally includes small usage examples to help demo how the API would be used_

Please don't forget to add new code usage examples to help understand how the modified API structure ties up together.

cc @Jozkee

**Comment by iSazonov** [comment]
@carlossanlop Thanks for clarification! My today priority is the recursion depth limit, it blocks me. The recursion depth limit API proposal doesn't seem to correlate with other proposed APIs so maybe move it in new issue to move forward faster?

**Comment by carlossanlop** [comment]
If you don't think they depend will each other, then sure, then I don't mind if we address that API separately.

But for simplicity and clarity, since the other APIs are not ready yet, can you please move the MaxDepthRecursion API to a separate API Proposal issue?

**Comment by iSazonov** [comment]
@carlossanlop I put new proposal to #43748.

**Comment by iSazonov** [comment]
>         public delegate bool ContinueOnErrorPredicate(ref FileSystemEntry parent, ReadOnlySpan<char> entry, int error);

I have a concern about `parent` parameter. It seems it will be difficult to implement at subdirectory dequeuing - current implementation does not track parent entry/path and adding this will decrease performance significantly.

As for generalization error reporting, have we any samples of such maooings in the repo?
We could create an exception object like UnauthorizedAccessException/IOException and its.

**Comment by Jozkee** [comment]
@iSazonov Can we remove `MaxDepthRecursion` from this proposal since you already created a separate one for it?

**Comment by iSazonov** [comment]
@Jozkee Done.

**Comment by iSazonov** [comment]
It seems this is waiting me :-)

If we look current implementation we can get only two **actual** entities in error time - (1) native error number, (2) current path (on which the error is raised), and we haven't FileSystemEntry in the time.

So proposal is `struct FileSystemErrorInfo` to wrap these two values. Also we would want to get an OS independent result. Again, current implementation already throws an OS independent exception generated by internal method ` Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(result), _currentPath, isDirectory: true);` on Unix and `Win32Marshal.GetExceptionForWin32Error(error, _currentPath)`. So we can expose the methods in FileSystemErrorInfo with GetException(). This is great to have in PowerShell where we can not only _throw_ but also write a _warning_ instead of the throw. (Ex.: if we do `dir -Recuse`, PowerShell writes warning messages to user on access denied errors and _continue_ enumeration.

```c#
namespace System.IO.Enumeration
{ 
    public class FileSystemEnumerable<TResult> : IEnumerable<TResult>
    {
        public FindPredicate? ShouldIncludePredicate { get; set; } // Existing API
        public FindPredicate? ShouldRecursePredicate { get; set; } // Existing API

        public delegate void OnDirectoryFinishedDelegate(ref FileSystemEntry entry); // New API
        public OnDirectoryFinishedDelegate? OnDirectoryFinishedAction { get; set; } // New API


        public delegate bool ContinueOnErrorPredicate(FileSystemErrorInfo errorInfo); // New API
        public ContinueOnErrorPredicate? ShouldContinueOnErrorPredicate { get; set; } // New API
    }

    public readonly ref struct FileSystemErrorInfo
    {
        // internal FileSystemErrorInfo(int errno, string path);
        public int Error { get; }; // 'errno' is converted on Unix with `internal struct ErrorInfo` to platform-agnostic Error
        public string Path { get; };
        public Exception GetException(); // Gets platform-agnostic exception for the error
        public string GetErrorMessage(); // Optional for me but could be useful due to absence of extra allocations with GetException()
    }

    public unsafe abstract partial class FileSystemEnumerator<TResult> : CriticalFinalizerObject, IEnumerator<TResult>
    {
        [Obsolete("The method is deprecated. Use ContinueOnError(FileSystemErrorInfo errorInfo) instead.")] // New
        protected virtual bool ContinueOnError(int error); // Deprecate
        protected virtual bool ContinueOnError(FileSystemErrorInfo errorInfo);  New API
    }
}
```

**Comment by iSazonov** [comment]
@danmoseley @Jozkee Friendly ping.

**Comment by iSazonov** [comment]
I'd happy to implement this for PowerShell.

**Comment by Jozkee** [comment]
For `OnDirectoryFinishedDelegate(ref FileSystemEntry entry)`: do we want to add an equivalent API to `FileSystemEnumerator` and obsolete the current one taking just a ReadOnlySpan<char>?

For `bool ContinueOnErrorPredicate(FileSystemErrorInfo errorInfo)`: can we consider not adding another type and instead re-use FileSystemEntry? If so, I think the API can look like `bool ContinueOnErrorPredicate(int error, ref FileSystemEntry entry)` I don't see very useful `GetException()` and `GetErrorMessage()`.


@iSazonov if you are blocked by this, you think you can extend the Enumerator and work with that? e.g:

```cs
class MyFileSystmeEnumerator : FileSystemEnumerator<string>
{
    public MyFileSystmeEnumerator(string directory, EnumerationOptions? options = null) : base(directory, options) { }

    protected override string TransformEntry(ref FileSystemEntry entry)
    {
        return entry.ToFullPath();
    }

    protected override void OnDirectoryFinished(ReadOnlySpan<char> directory)
    {
        // Do what you need
    }

    protected override bool ContinueOnError(int error)
    {
        // Do what you need
    }
}
```

**Comment by iSazonov** [comment]
>For bool ContinueOnErrorPredicate(FileSystemErrorInfo errorInfo): can we consider not adding another type and instead re-use FileSystemEntry? If so, I think the API can look like bool ContinueOnErrorPredicate(int error, ref FileSystemEntry entry) I don't see very useful GetException() and GetErrorMessage().

If we look current implementation there is no FileSystemEntry instance in points where the error is detected. And I don't see needs to create the instance for "parent" path in the error report (user himself can create Directory/FileInfo if needed) - we can easily have parent path but what about error code - it will OS specific but we need OS agnostic.

As for GetException() and GetErrorMessage(), these address previous API review request to have OS agnostic error report since today we get an _OS specific Int value_. We could remove FileSystemErrorInfo from the API proposal and return Exception which contains the native error code, but we lost parent path since it will be in error message only. So we again come back to previous API review and suggestion to have new (extendable) type for error reporting.

>For OnDirectoryFinishedDelegate(ref FileSystemEntry entry): do we want to add an equivalent API to FileSystemEnumerator and obsolete the current one taking just a ReadOnlySpan?

Yes, it makes sense.

>if you are blocked by this, you think you can extend the Enumerator and work with that? e.g:

I can not process ContinueOnError(int error) well and as result the issue was opened :-)

**Comment by Jozkee** [comment]
> these address previous API review request to have OS agnostic 

That's interesting because it contradicts the current `FileSystemEnumerator<TResult>.ContinueOnError(Int32)` API which was approved by the same board. Did you get that feedback for this specific proposal? If not, I think this can be an exception and can be discussed live.

**Comment by iSazonov** [comment]
> That's interesting because it contradicts the current FileSystemEnumerator<TResult>.ContinueOnError(Int32) API which was approved by the same board. Did you get that feedback for this specific proposal? If not, I think this can be an exception and can be discussed live.

Sorry, it seems I don't understand your question. If I remember right Jeremy made a proposal for file enumeration with a lot of new APIs, some from them (including ContinueOnError) was under "need feedback from community". First feedback (as [Jeremy mentioned](https://github.com/dotnet/runtime/issues/1953#issuecomment-594948819)) was community want delegates for all virtual methods so I added OnDirectoryFinishedDelegate in the proposal. 
Second feedback comes from my experience with PowerShell code. As I already described PowerShell writes warning messages on (access denied) errors and continue enumeration. (And I'd happy to avoid OS Specific error code processing.) So I added new ContinueOnErrorPredicate in the proposal. 

> Can you provide samples that will be addressed by adding this new set of APIs?

In my old PR https://github.com/PowerShell/PowerShell/pull/12834 you can see how weird looks code (notice, I don't OS specific error code processing there for the example simplicity!) with ContinueOnError(int error) - with ContinueOnError(FileSystemErrorInfo errorInfo) we get exactly we need.

**Comment by Jozkee** [comment]
> First feedback (as https://github.com/dotnet/runtime/issues/1953#issuecomment-594948819) was community want delegates for all virtual methods so I added OnDirectoryFinishedDelegate in the proposal.

I agree this proposal is about adding delegates for all virtual in the Enumerator and set them as properties in the Enumerable. What I am pointing out is that the virtual method `FileSystemEnumerator<TResult>.ContinueOnError(Int32)` already sets precedence for receiving [the native error code](https://docs.microsoft.com/en-us/dotnet/api/system.io.enumeration.filesystemenumerator-1.continueonerror?view=net-6.0#parameters) and should consider follow that precedence. 

So I would suggest having FileSystemErrorInfo as an alternative design.

> In my old PR https://github.com/PowerShell/PowerShell/pull/12834 you can see how weird looks code (notice, I don't OS specific error code processing there for the example simplicity!) with ContinueOnError(int error) - with ContinueOnError(FileSystemErrorInfo errorInfo) we get exactly we need.

Could you please write a minimal sample in here? It would be hard to scroll through your PR while reviewing the proposal.

**Comment by iSazonov** [comment]
> Could you please write a minimal sample in here? It would be hard to scroll through your PR while reviewing the proposal.

Yes, of cause.
```c#
using System.IO.Enumeration;

string directory = "...";
var context = ... // PowerShell context

var files = new FileSystemEnumerable<string>(
    directory,
    // Yes, it is weird but we should write right error for root directory too.
    (ref FileSystemEntry entry) => entry.OriginalRootDirectory.Length > entry.Directory.Length ? entry.FileName.ToString() : Path.Join(entry.Directory.Slice(entry.OriginalRootDirectory.Length), entry.FileName),
    options)
{
    ShouldIncludePredicate = (ref FileSystemEntry entry) => FileSystemEntryFilter, // PowerSHell include filter delegate
    ShouldRecursePredicate = (ref FileSystemEntry entry) =>
    {
        if (context.Recurce && context.IsFollowSymlink && entry.Attributes.HasFlag(FileAttributes.ReparsePoint) && InternalSymbolicLinkLinkCodeMethods.IsReparsePointWithTarget(entry.ToFileSystemInfo()))
        {
            // Follow symlinks at recursion.
            return true;
        }

        return context.Recurce;
    },
    ShouldContinueOnErrorPredicate = (FileSystemErrorInfo errorInfo) =>
    {
        var exc = errorInfo.GetException();
        if (exc is UnauthorizedAccessException)
        {
            context.WriteError(new ErrorRecord(exc, "DirUnauthorizedAccessError", ErrorCategory.PermissionDenied, errorInfo.Path));
        }
        else
        {
            context.WriteError(new ErrorRecord(exc, "DirIOError", ErrorCategory.ReadError, errorInfo.Path));
        }

        return true;
    }
};
```

**Comment by bartonjs** [comment]
* There's no need for ContinueOnErrorPredicate, it's `Func<int, bool>`
* OnDirectoryFinishedAction should use System.String, not `ReadOnlySpan<char>`.  And then it becomes `Action<string>`.
* We added OnDirectoryStartedAction
* We then removed the prefix "On", but left the suffix "Action" to match the existing "-Predicate" properties.
* We aren't sure if we should remove the "Should" prefix in the on-error handler, or if it's helpful.  

**Action Required:**
* We questioned if the error handler predicate should also take the path that failed.
  * And we didn't know what it should look like (`Func<string, int, bool>`, something complicated with current directory and current candidate as different parameters, etc)


```C#
public class FileSystemEnumerable<TResult> : IEnumerable<TResult>
{
    public Action<string>? DirectoryStartedAction { get; set; } // New API
    public Action<string>? DirectoryFinishedAction { get; set; } // New API

    public Func<int, bool>? ShouldContinueOnErrorPredicate { get; set; } // New API
}
```

**Comment by iSazonov** [comment]
I agree it is important to pass current context to custom error handler.
[Previous proposal]() contains `readonly ref struct FileSystemErrorInfo` to pass a context to the custom error handler.
But if we would agree errors in enumaration time are rare we could pass exception instance which contains all information we need and more (like resource string message). This is very convenient and eliminates the need to duplicate non-trivial code `Interop.GetExceptionForIoErrno()`. Notice, the exception is OS indepentent!
So we could rewrite follow https://github.com/dotnet/runtime/blob/2f3fcca06499725469b6ae021be960e08d808c75/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs#L69-L82
with
```c#
        private IntPtr CreateDirectoryHandle(string path, bool ignoreNotFound = false)
        {
            IntPtr handle = Interop.Sys.OpenDir(path);
            if (handle == IntPtr.Zero)
            {
                Interop.ErrorInfo info = Interop.Sys.GetLastErrorInfo();
                if (InternalContinueOnError(info, ignoreNotFound))
                {
                    return IntPtr.Zero;
                }

		var exception = Interop.GetExceptionForIoErrno(info, path, isDirError: true)
                if (ContinueOnError(exception))
                {
                    return IntPtr.Zero;
                }
                throw exception;
            }
            return handle;
        }
```

This is consistent with the way PowerShell works - catch an exception and print a friendly message to the user, i.e. the exception instance is used anyway.
If there are applications where efficiency is critical at this point, they can use the existing `ContinueOnError(int error)`.

Current proposal:
```c#
namespace System.IO.Enumeration
{ 
    public class FileSystemEnumerable<TResult> : IEnumerable<TResult>
    {
        public FindPredicate? ShouldIncludePredicate { get; set; } // Existing API
        public FindPredicate? ShouldRecursePredicate { get; set; } // Existing API

        public Action<string>? DirectoryStartedAction { get; set; } // New API
        public Action<string>? DirectoryFinishedAction { get; set; } // New API

        public Func<Exception, bool>? ShouldContinueOnErrorPredicate { get; set; } // New API
    }

    public unsafe abstract partial class FileSystemEnumerator<TResult> : CriticalFinalizerObject, IEnumerator<TResult>
    {
        [Obsolete("The method is deprecated. Use ContinueOnError(FileSystemErrorInfo errorInfo) instead.")] // New API
        protected virtual bool ContinueOnError(int error); // Deprecate or no?
        protected virtual bool ContinueOnError(Exception exception);  // New API
    }
}
```

**Comment by CyberSinh** [comment]
I understand that this thread still needs some work before it can be reviewed again. 
What exactly is expected from the .NET team? 
No feedback has been given on the API suggestion proposed by @iSazonov above.
